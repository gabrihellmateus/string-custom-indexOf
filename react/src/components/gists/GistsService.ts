import {GistFilesModel, GistDOMModel, GistModel, GistsResponseModel} from './GistsModel';
import {settings} from '../../settings';

export const GistsService = {
  getUrl: (params = settings.github.urlParams): string => `${params.url}/${params.user}/gists?page=1&per_page=${params.limit}`,

  getGists: async function(): Promise<GistsResponseModel> {
    const url = this.getUrl();

    try {
      const response = await fetch(url)
      return await response.json()
    } catch (error) {
      return error
    }
  },

  mapGists: async function(): Promise<GistDOMModel[]> {
    const gists: GistsResponseModel = await this.getGists();
    
    return Object.values(gists).map((gist: GistModel) => {
      const files: GistFilesModel[] = Object.values(gist.files);

      return {
        id: gist.id,
        url: gist.html_url,
        files: files,
        title: gist.description || settings.gist.title,
        language: files[0].language || settings.gist.logo
      };
    });
  }
};
