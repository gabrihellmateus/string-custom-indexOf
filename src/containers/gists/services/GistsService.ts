import {GistFilesModel, GistDOMModel, GistModel, GistsData} from '../models'
import {settings} from '../../../settings'
import {isDevelopment} from '../../../helpers'
import StorageService from '../../../shared/services/StorageService'

const storageService = StorageService()

export const GistsService = {
  headers: {
    Authorization: `token ${process.env.REACT_APP_GH_TOKEN}`
  },

  getUrl: (params = settings.github.urlParams): string =>
    `${params.url}/${params.user}/gists?per_page=${params.perPage}`,

  fetchGists: async function (): Promise<GistModel[]> {
    const url = this.getUrl()
    const headers = {
      ...(isDevelopment() && {headers: {...this.headers}})
    }

    try {
      const response = await fetch(url, headers)
      return await response.json()
    } catch (error) {
      return error
    }
  },

  mapGists: function (gists: GistModel[]): GistDOMModel[] {
    return gists.map((gist: GistModel) => {
      const files: GistFilesModel[] = Object.values(gist.files)
      const title = gist.description.replace(settings.gists.regexLogo, '').trim()
      const languageMatch = gist.description.match(settings.gists.regexLogo)
      const language =
        languageMatch && languageMatch[1] && languageMatch[1].replace(settings.gists.regexWebsite, '').trim()

      return {
        id: gist.id,
        url: gist.html_url,
        files: files,
        title: title || settings.gists.title,
        language: language || files[0].language || settings.gists.logo
      }
    })
  },

  filterGists: function (gists: GistModel[]): GistModel[] {
    const regex = new RegExp(settings.gists.regexWebsite, 'gmi')

    return gists
      .filter((gist) => gist.public)
      .filter((gist: GistModel) => gist.description.match(regex))
      .slice(0, settings.gists.limit)
      .map((gist: GistModel) => ({
        ...gist,
        description: gist.description.replace(regex, '').trim()
      }))
  },

  getGists: async function (): Promise<GistDOMModel[]> {
    const gists = await this.fetchGists()
    const filteredGists = this.filterGists(gists)
    console.log(filteredGists)
    const mappedGists = this.mapGists(filteredGists)

    return mappedGists
  },

  setGistsLocalStorage: (data: GistsData) => {
    if (data) {
      storageService.setItem('gists', JSON.stringify(data))
    }
  },

  getGistsLocalStorage: (): GistsData => {
    const gists = storageService.getItem('gists')

    return gists ? JSON.parse(gists) : null
  },

  shouldSetGistsLocalStorage: (savedGists: GistsData): boolean => {
    if (!savedGists || !savedGists.date) return false

    const hour = 1000 * 60 * 60
    const diff = (Date.now() - savedGists.date) / hour

    return Math.abs(Math.round(diff)) > 1
  }
}
