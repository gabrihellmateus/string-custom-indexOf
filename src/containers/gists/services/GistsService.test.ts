import fetchMock, {enableFetchMocks} from 'jest-fetch-mock'

import {GistsService} from './GistsService'

import {gistsResponseMock} from '../../../../test/mocks'

describe('GistsService', () => {
  beforeEach(() => {
    Object.defineProperty(window, 'localStorage', {
      value: {
        getItem: jest.fn(() => JSON.stringify(gistsResponseMock)),
        setItem: jest.fn(() => gistsResponseMock)
      },
      writable: true
    })

    enableFetchMocks()
  })

  afterEach(() => {
    fetchMock.resetMocks()
  })

  describe('Fetching Gists', () => {
    it('fetchGists: return fetched gists', async (done) => {
      fetchMock.mockResponseOnce(JSON.stringify(gistsResponseMock))

      const result = await GistsService.fetchGists()

      expect(result).toMatchSnapshot()
      done()
    })

    it('fetchGists: return error when trying to fetch gists', async (done) => {
      fetchMock.mockRejectOnce(new Error('Not Found!'))
      jest.spyOn(GistsService, 'getUrl').mockImplementationOnce(() => 'https://example.com')

      const result = await GistsService.fetchGists()

      expect(result.toString()).toBe('Error: Not Found!')
      expect((fetch as any).mock.calls.length).toEqual(1)
      expect((fetch as any).mock.calls[0][0]).toEqual('https://example.com')
      expect(result).toMatchSnapshot()
      done()
    })
  })

  describe('Data', () => {
    it('getUrl: return an url', () => {
      const urlParam = {
        url: 'www.example.com',
        user: 'john-doe',
        perPage: 100
      }

      const result = 'www.example.com/john-doe/gists?per_page=100'

      expect(GistsService.getUrl(urlParam)).toBe(result)
    })

    it('mapGists: return a mapped array of gists', () => {
      const result = GistsService.mapGists(gistsResponseMock as any)

      expect(result).toMatchSnapshot()
    })
  })

  describe('Local Storage', () => {
    it('setGistsLocalStorage: should set gists in local storage', () => {
      Object.defineProperty(window, 'localStorage', {
        value: {
          setItem: jest.fn(() => gistsResponseMock)
        }
      })

      const now = Date.now()
      GistsService.setGistsLocalStorage({
        date: now,
        collection: [
          {
            id: '1',
            url: 'https://example.com',
            files: [
              {
                filename: 'test.js',
                type: 'application/javascript',
                language: 'JavaScript',
                raw_url: 'https://example.com/gist',
                size: 1000
              }
            ],
            title: 'Gist Sample',
            language: 'JavaScript'
          }
        ]
      })

      expect(window.localStorage.setItem).toHaveBeenCalledWith(
        'GABRIEL-LIMA:GISTS',
        `{"date":${now},"collection":[{"id":"1","url":"https://example.com","files":[{"filename":"test.js","type":"application/javascript","language":"JavaScript","raw_url":"https://example.com/gist","size":1000}],"title":"Gist Sample","language":"JavaScript"}]}`
      )
    })

    it('setGistsLocalStorage: should not call localStorage setItem', () => {
      Object.defineProperty(window, 'localStorage', {
        value: {
          setItem: jest.fn(() => gistsResponseMock)
        }
      })

      GistsService.setGistsLocalStorage(null as any)

      expect(window.localStorage.setItem).toHaveBeenCalledTimes(0)
    })

    it('getGistsLocalStorage: should get gists from local storage', () => {
      Object.defineProperty(window, 'localStorage', {
        value: {
          getItem: jest.fn(() => JSON.stringify(gistsResponseMock))
        }
      })

      const result = GistsService.getGistsLocalStorage()

      expect(window.localStorage.getItem).toHaveBeenCalledWith('GABRIEL-LIMA:GISTS')
      expect(result).toMatchSnapshot()
    })

    it('getGistsLocalStorage: should return null when can not find gists from local storage', () => {
      Object.defineProperty(window, 'localStorage', {
        value: {
          getItem: jest.fn(() => null)
        }
      })

      const result = GistsService.getGistsLocalStorage()

      expect(window.localStorage.getItem).toHaveBeenCalledWith('GABRIEL-LIMA:GISTS')
      expect(result).toBeNull()
    })
  })
})
