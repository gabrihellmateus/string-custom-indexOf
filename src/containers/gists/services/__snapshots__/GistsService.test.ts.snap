// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GistsService getGistsLocalStorage: should get gists from local storage 1`] = `
Object {
  "data": Object {
    "collection": Array [
      Object {
        "files": Array [
          Object {
            "filename": ".hyper.js",
            "language": "JavaScript",
            "raw_url": "https://gist.githubusercontent.com/the-glima/2980bf771357f4caab71e8ae571f10c9/raw/e78cec2e2086dab2e96e766b401d08312a6e67a2/.hyper.js",
            "size": 5898,
            "type": "application/javascript",
          },
        ],
        "id": "1",
        "language": "JavaScript",
        "title": "Hyper Configuration",
        "url": "https://gist.github.com/2980bf771357f4caab71e8ae571f10c9",
      },
      Object {
        "files": Array [
          Object {
            "filename": "get-latest-tag-on-git.sh",
            "language": "Shell",
            "raw_url": "https://gist.githubusercontent.com/the-glima/5b5503c60df6371c22d611a0c87c2821/raw/486a95da7c14fbd3c5cc7801f273bf72a9a28ea8/get-latest-tag-on-git.sh",
            "size": 637,
            "type": "application/x-sh",
          },
        ],
        "id": "5b5503c60df6371c22d611a0c87c2821",
        "language": "Shell",
        "title": "Getting latest tag on git repository",
        "url": "https://gist.github.com/5b5503c60df6371c22d611a0c87c2821",
      },
      Object {
        "files": Array [
          Object {
            "filename": "git-show-commits-range.sh",
            "language": "Shell",
            "raw_url": "https://gist.githubusercontent.com/the-glima/e19892f44976495b43b68cbf075d0d1b/raw/ab907b151ddb09033cf5131dae455e75bb19047b/git-show-commits-range.sh",
            "size": 308,
            "type": "application/x-sh",
          },
        ],
        "id": "e19892f44976495b43b68cbf075d0d1b",
        "language": "Shell",
        "title": "Show commits by a range of time",
        "url": "https://gist.github.com/e19892f44976495b43b68cbf075d0d1b",
      },
      Object {
        "files": Array [
          Object {
            "filename": "sync-branch.sh",
            "language": "Shell",
            "raw_url": "https://gist.githubusercontent.com/the-glima/5912b8c98bef97c19bc4c0ca3b4e29ee/raw/dbcd7224a8d8ca6cef5412b106ad71c2cab3eb4a/sync-branch.sh",
            "size": 579,
            "type": "application/x-sh",
          },
        ],
        "id": "5912b8c98bef97c19bc4c0ca3b4e29ee",
        "language": "Shell",
        "title": "Sync your current branch with another",
        "url": "https://gist.github.com/5912b8c98bef97c19bc4c0ca3b4e29ee",
      },
      Object {
        "files": Array [
          Object {
            "filename": "multiple_ssh_setting.md",
            "language": "Markdown",
            "raw_url": "https://gist.githubusercontent.com/the-glima/37d90c5060f8d9dd2e5bdb6611c416f5/raw/58a7c960d1fddc642b6d60e76633210cd4b659ec/multiple_ssh_setting.md",
            "size": 1797,
            "type": "text/markdown",
          },
        ],
        "id": "37d90c5060f8d9dd2e5bdb6611c416f5",
        "language": "Markdown",
        "title": "Multiple SSH keys GitHub setup",
        "url": "https://gist.github.com/37d90c5060f8d9dd2e5bdb6611c416f5",
      },
      Object {
        "files": Array [
          Object {
            "filename": "pr-check.js",
            "language": "JavaScript",
            "raw_url": "https://gist.githubusercontent.com/the-glima/86e7eec2a597ee7a22121b11cab32a0d/raw/75e59e6ab4818b287451556a7a7b76692951ef85/pr-check.js",
            "size": 3930,
            "type": "application/javascript",
          },
        ],
        "id": "86e7eec2a597ee7a22121b11cab32a0d",
        "language": "JavaScript",
        "title": "Automated checks when creating PRs",
        "url": "https://gist.github.com/86e7eec2a597ee7a22121b11cab32a0d",
      },
      Object {
        "files": Array [
          Object {
            "filename": "string.customIndexOf.js",
            "language": "JavaScript",
            "raw_url": "https://gist.githubusercontent.com/the-glima/9ca6a248e61c3ed74859e30f0952bf3c/raw/71f20b957ce4c74bc329076876456dbb9a6508fc/string.customIndexOf.js",
            "size": 7472,
            "type": "application/javascript",
          },
        ],
        "id": "9ca6a248e61c3ed74859e30f0952bf3c",
        "language": "JavaScript",
        "title": "Custom String.prototype.indexOf",
        "url": "https://gist.github.com/9ca6a248e61c3ed74859e30f0952bf3c",
      },
      Object {
        "files": Array [
          Object {
            "filename": "prepare-commit-msg.sh",
            "language": "Shell",
            "raw_url": "https://gist.githubusercontent.com/the-glima/ccca711a04e5e9951b1a/raw/5929054f65dbefbf210b2f5eb399e1e344310ef5/prepare-commit-msg.sh",
            "size": 549,
            "type": "application/x-sh",
          },
        ],
        "id": "ccca711a04e5e9951b1a",
        "language": "Shell",
        "title": "Prepend branch name into commit message",
        "url": "https://gist.github.com/ccca711a04e5e9951b1a",
      },
    ],
    "date": 1000000,
  },
}
`;
